src/gluonts/zebras/_timeframe.py

```diff
@@ -0,0 +1,305 @@
+        pad_left = left + self._pad.left
+        pad_right = right + self._pad.right
+
+        index = self.index
+        if index is not None:
+            index = self.index.prepend(left).extend(right)
+
+        return _replace(
+            self,
+            columns=columns,
+            index=index,
+            length=length,
+            _pad=(pad_left, pad_right),
+        )
+
+    def index_of(self, period: Period):
+        assert self.index is not None
+
+        return self.index.index_of(period)
+
+    def astype(self, type):
+        return _replace(
+            self,
+            columns=valmap(type, self.columns),
+            static=valmap(type, self.static),
+        )
+
+    def __len__(self):
+        return self.length
+
+    def __repr__(self):
+        columns = ", ".join(self.columns)
+        return f"TimeFrame<size={len(self)}, columns=[{columns}]>"
+
+    def _repr_html_(self):
+        columns = {}
+
+        if self.index is not None:
+            columns[""] = pluck_attr(self.index, "data")
+
+        columns.update(self.columns)
+
+        table = html_table(columns)
+        return table + f"{len(self)} rows × {len(self.columns)} columns"
+
+    def set(self, name, value, tdim=None):
+        tdim = maybe.unwrap_or(tdim, self.default_tdim)
+
+        return _replace(
+            columns=merge(self.columns, {name: value}),
+            tdims=merge(self.tdims, {name: tdim}),
+        )
+
+    def remove(self, column):
+        return _replace(
+            self,
+            columns=dissoc(self.columns, column),
+            tdims=dissoc(self.tdims, column),
+        )
+
+    def stack(
+        self,
+        columns: List[str],
+        into: str,
+        drop: bool = True,
+    ) -> TimeFrame:
+        # Ensure all tdims are the same.
+        # TODO: Can we make that work for different tdims? There might be a
+        # problem with what the resulting dimensions are.
+        assert len(set([self.tdims[column] for column in columns])) == 1
+
+        if drop:
+            columns = dissoc(self.columns, *columns)
+            tdims = dissoc(self.tdims, *columns)
+        else:
+            columns = self.columns
+            tdims = self.tdims
+
+        columns[into] = np.vstack([self.columns[column] for column in columns])
+
+        return _replace(self, columns=columns, tdims=tdims)
+
+    def with_index(self, index):
+        return _replace(self, index=index)
+
+    def as_dict(self, prefix=None, pad=None, static=True):
+        result = dict(self.columns)
+
+        if prefix is not None:
+            result = {prefix + key: value for key, value in result.items()}
+
+        if static:
+            result.update(self.static)
+
+        if pad is not None:
+            result[pad] = self._pad
+
+        return result
+
+
+def time_frame(
+    columns=None,
+    index=None,
+    start=None,
+    freq=None,
+    static=None,
+    tdims=None,
+    length=None,
```

Comment:
I'm wondering when would one need to specify this?

Reply:
If you create an empty instance which you want to build up.

Reply:
Wouldn't one in that case provide a non-empty set of columns, each one being empty?

Reply:
Not sure what you mean, but what happens if you drop all columns -- the length should stay the same.